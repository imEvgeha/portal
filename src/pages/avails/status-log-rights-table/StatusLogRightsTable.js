import React, {useEffect, useState} from 'react';
import PropTypes from 'prop-types';
import {getUsername} from "@portal/portal-auth/authSelectors";
import NexusDrawer from '@vubiquity-nexus/portal-ui/lib/elements/nexus-drawer/NexusDrawer';
import NexusGrid from '@vubiquity-nexus/portal-ui/lib/elements/nexus-grid/NexusGrid';
import {GRID_EVENTS} from '@vubiquity-nexus/portal-ui/lib/elements/nexus-grid/constants';
import {defineCheckboxSelectionColumn} from '@vubiquity-nexus/portal-ui/lib/elements/nexus-grid/elements/columnDefinitions';
import withColumnsResizing from '@vubiquity-nexus/portal-ui/lib/elements/nexus-grid/hoc/withColumnsResizing';
import withFilterableColumns from '@vubiquity-nexus/portal-ui/lib/elements/nexus-grid/hoc/withFilterableColumns';
import withInfiniteScrolling from '@vubiquity-nexus/portal-ui/lib/elements/nexus-grid/hoc/withInfiniteScrolling';
import withSideBar from '@vubiquity-nexus/portal-ui/lib/elements/nexus-grid/hoc/withSideBar';
import {connect} from 'react-redux';
import {compose} from 'redux';
import {ERROR_TABLE_COLUMNS, ERROR_TABLE_TITLE} from '../../sync-log/syncLogConstants';
import AvailsTableToolbar from '../avails-table-toolbar/AvailsTableToolbar';
import {STATUS_TAB} from '../rights-repository/constants';
import {setPreplanRights} from '../rights-repository/rightsActions';
import * as selectors from '../rights-repository/rightsSelectors';
import StatusRightsActions from '../status-rights-actions/StatusRightsActions';
import {getStatusLog} from './StatusLogService';
import columnMappings from './columnMappings';
import './StatusLogRightsTable.scss';
import StatusLogErrors from './components/PublishErrors/StatusLogErrors';
import SelectedStatusTable from './components/selected-status-table/SelectedStatusTable';

const StatusLogRightsGrid = compose(
    withSideBar(),
    withColumnsResizing(),
    withFilterableColumns({frameworkComponents: {publishErrors: StatusLogErrors}}),
    withColumnsResizing(),
    withInfiniteScrolling({fetchData: getStatusLog})
)(NexusGrid);

export const StatusLogRightsTable = ({totalRowCount, username}) => {
    const [showDrawer, setShowDrawer] = useState(false);
    const [errorsData, setErrorsData] = useState([]);
    const [showSelected, setShowSelected] = useState(false);
    const [gridApi, setGridApi] = useState(undefined);
    const [columnApiState, setColumnApiState] = useState(undefined);
    const [selectedRights, setSelectedRights] = useState([]);
    const [columnDefs, setColumnDefs] = useState([]);

    const closeDrawer = () => setShowDrawer(false);

    useEffect(() => {
        const checkBoxColDef = defineCheckboxSelectionColumn({
            headerCheckboxSelection: true,
            headerCheckboxSelectionFilteredOnly: true,
        });

        const mappings = columnMappings.map(col => ({...col, cellRendererParams: {setErrors}}));
        setColumnDefs([checkBoxColDef, ...mappings]);
    }, []);

    const setErrors = data => {
        setErrorsData(data);
        setShowDrawer(true);
    };

    const onGridEvent = ({type, api, columnApi}) => {
        const {READY, SELECTION_CHANGED} = GRID_EVENTS;
        switch (type) {
            case READY:
                !gridApi && setGridApi(api);
                !columnApiState && setColumnApiState(columnApi);
                api.sizeColumnsToFit();
                break;
            case SELECTION_CHANGED:
                setSelectedRights(api.getSelectedRows());
                break;
            default:
                break;
        }
    };

    const selectableRows = ['SUCCESS', 'DELETED'];
    const isRowSelectable = rowNode => rowNode?.data && !selectableRows.includes(rowNode?.data?.status);

    const toolbarActions = () => {
        return <StatusRightsActions rights={selectedRights} />;
    };

    return (
        <div className="nexus-c-status-log-table">
            <AvailsTableToolbar
                totalRecordsCount={totalRowCount}
                activeTab={STATUS_TAB}
                selectedRowsCount={selectedRights.length}
                setIsSelected={setShowSelected}
                isSelected={showSelected}
                gridApi={gridApi}
                columnApi={columnApiState}
                username={username}
                showSelectedButton={true}
                toolbarActions={toolbarActions()}
            />

            {!showSelected && (
                <StatusLogRightsGrid
                    suppressRowClickSelection
                    className="nexus-c-status-log-grid"
                    columnDefs={columnDefs}
                    mapping={columnMappings}
                    rowSelection="multiple"
                    onGridEvent={onGridEvent}
                    rowClassRules={{
                        'disable-selected': params => ['SUCCESS', 'DELETED'].includes(params?.data?.status),
                    }}
                    context={{selectedRows: [...selectedRights]}}
                    isRowSelectable={isRowSelectable}
                />
            )}

            {showSelected && (
                <SelectedStatusTable
                    columnDefs={columnDefs}
                    selectedStatusRights={selectedRights}
                    setSelectedStatusRights={setSelectedRights}
                />
            )}

            <NexusDrawer onClose={closeDrawer} isOpen={showDrawer} title={ERROR_TABLE_TITLE} width="wider">
                <div className="nexus-c-sync-log-table__errors-table">
                    {ERROR_TABLE_COLUMNS.map(column => (
                        <div className="nexus-c-sync-log-table__errors-table--header-cell" key={column}>
                            {column.toUpperCase()}
                        </div>
                    ))}
                    {errorsData.map((error, i) =>
                        ERROR_TABLE_COLUMNS.map(key => (
                            <div className="nexus-c-sync-log-table__errors-table--cell" key={`error-${i}-${key}`}>
                                {error.split(' - ')[key === 'type' ? 0 : 1]}
                            </div>
                        ))
                    )}
                </div>
            </NexusDrawer>
        </div>
    );
};

StatusLogRightsTable.propTypes = {
    totalRowCount: PropTypes.number,
    username: PropTypes.string.isRequired,
};

StatusLogRightsTable.defaultProps = {
    totalRowCount: 0,
};

const mapStateToProps = () => {
    const totalRowCountSelector = selectors.createStatusLogCountSelector();
    return (state, props) => ({
        totalRowCount: totalRowCountSelector(state, props),
        username: getUsername(state),
    });
};

const mapDispatchToProps = dispatch => ({
    setPreplanRights: payload => dispatch(setPreplanRights(payload)),
});
export default connect(mapStateToProps, mapDispatchToProps)(StatusLogRightsTable);
