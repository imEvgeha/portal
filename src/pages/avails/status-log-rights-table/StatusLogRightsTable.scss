@import '~@vubiquity-nexus/portal-styles/scss/atlaskit/colors';
@import '~@vubiquity-nexus/portal-styles/scss/config';
@import '~@vubiquity-nexus/portal-styles/scss/atlaskit/mixins';

$border: 1px solid rgba(map-get($colors, nexus-title-color), 0.7);

.nexus-c-status-log-table {
    @include padding(0, 0, 10, 0);
    height: 100%;
    display: flex;
    flex-direction: column;
    &__errors-table {
        margin: 20px;
        display: grid;
        grid-template-rows: 2em repeat(auto-fill, minmax(0, auto));
        grid-template-columns: repeat(2, minmax(0, auto));
        border: $border;
        border-right: none;
        &--header-cell {
            width: 100%;
            display: flex;
            justify-content: flex-start;
            color: map-get($colors, nexus-title-color);
            text-align: left;
            border-bottom: $border;
            border-right: $border;
            padding: 5px;
            font-weight: bold;
        }

        &--cell {
            padding: 5px;
            border-right: $border;
            border-bottom: $border;
            background: rgba(map_get($colors, nexus-title-color), 0.1);

            // fixes content leaking from all table cells
            // see http://agile.vubiquity.com/browse/PORT-2513
            overflow-wrap: break-word;
            word-wrap: break-word;
            word-break: break-all;
            hyphens: auto;
        }
    }
    &__error-cell {
        text-decoration: underline;
        color: map-get($colors, nexus-link);
        cursor: pointer;
    }

    .ag-cell {
        text-overflow: ellipsis;
        overflow: hidden;
    }
}

.nexus-c-status-log-table__actions {
    display: flex;
    width: 100%;
    justify-content: space-between;
    align-items: flex-end;
}
