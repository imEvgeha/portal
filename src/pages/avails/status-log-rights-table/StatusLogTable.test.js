import React from 'react';
import {shallow} from 'enzyme';
import {StatusLogRightsTable} from './StatusLogRightsTable';

describe('StatusLogRightsTable', () => {
    let wrapper = null;
    const props = {
        activeTab: 'Status',
    };

    beforeEach(() => {
        wrapper = shallow(<StatusLogRightsTable {...props} />);
    });

    it('should match snapshot', () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('should render StatusLogRights table wrapper', () => {
        expect(wrapper.find('.nexus-c-sync-log-table').length).toEqual(1);
    });

    it('should render NexusDrawer', () => {
        expect(wrapper.find('NexusDrawer').length).toEqual(1);
    });
});
