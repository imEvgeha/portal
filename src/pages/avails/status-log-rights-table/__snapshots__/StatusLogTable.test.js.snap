// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatusLogRightsTable should match snapshot 1`] = `
<div
  className="nexus-c-status-log-table"
>
  <Component
    className="nexus-c-status-log-grid"
    columnDefs={
      Array [
        Object {
          "cellRenderer": "status",
          "cellRendererParams": Object {
            "setErrors": [Function],
          },
          "colId": "status",
          "dataType": "string",
          "enableSearch": true,
          "field": "status",
          "headerName": "Status",
          "javaVariableName": "status",
          "minWidth": 100,
          "searchDataType": "string",
        },
        Object {
          "cellRenderer": "titleNameCellRenderer",
          "cellRendererParams": Object {
            "setErrors": [Function],
          },
          "colId": "rightTitle",
          "field": "rightTitle",
          "headerName": "Right Title",
          "javaVariableName": "rightTitle",
          "minWidth": 150,
        },
        Object {
          "cellRendererParams": Object {
            "setErrors": [Function],
          },
          "colId": "entityId",
          "dataType": "string",
          "enableSearch": true,
          "field": "entityId",
          "headerName": "Right ID",
          "javaVariableName": "entityId",
          "minWidth": 150,
          "searchDataType": "string",
        },
        Object {
          "cellRendererParams": Object {
            "setErrors": [Function],
          },
          "colId": "titleId",
          "dataType": "string",
          "enableSearch": true,
          "field": "titleId",
          "headerName": "Title ID",
          "javaVariableName": "titleId",
          "minWidth": 150,
          "searchDataType": "string",
        },
        Object {
          "cellRendererParams": Object {
            "setErrors": [Function],
          },
          "colId": "externalSystem",
          "field": "externalSystem",
          "headerName": "External System",
          "javaVariableName": "externalSystem",
          "minWidth": 120,
        },
        Object {
          "cellRendererParams": Object {
            "setErrors": [Function],
          },
          "colId": "publishedAt",
          "dataType": "timestamp",
          "enableSearch": true,
          "field": "publishedAt",
          "headerName": "Publish date",
          "javaVariableName": "publishedAt",
          "minWidth": 150,
          "searchDataType": "regionalMidnight",
        },
        Object {
          "cellRenderer": "publishErrors",
          "cellRendererParams": Object {
            "setErrors": [Function],
          },
          "colId": "publishErrors",
          "field": "publishErrors",
          "headerName": "Errors",
          "javaVariableName": "publishErrors",
          "minWidth": 150,
        },
      ]
    }
    frameworkComponents={
      Object {
        "publishErrors": [Function],
      }
    }
    isGridHidden={false}
    mapping={
      Array [
        Object {
          "cellRenderer": "status",
          "colId": "status",
          "dataType": "string",
          "enableSearch": true,
          "field": "status",
          "headerName": "Status",
          "javaVariableName": "status",
          "minWidth": 100,
          "searchDataType": "string",
        },
        Object {
          "cellRenderer": "titleNameCellRenderer",
          "colId": "rightTitle",
          "field": "rightTitle",
          "headerName": "Right Title",
          "javaVariableName": "rightTitle",
          "minWidth": 150,
        },
        Object {
          "colId": "entityId",
          "dataType": "string",
          "enableSearch": true,
          "field": "entityId",
          "headerName": "Right ID",
          "javaVariableName": "entityId",
          "minWidth": 150,
          "searchDataType": "string",
        },
        Object {
          "colId": "titleId",
          "dataType": "string",
          "enableSearch": true,
          "field": "titleId",
          "headerName": "Title ID",
          "javaVariableName": "titleId",
          "minWidth": 150,
          "searchDataType": "string",
        },
        Object {
          "colId": "externalSystem",
          "field": "externalSystem",
          "headerName": "External System",
          "javaVariableName": "externalSystem",
          "minWidth": 120,
        },
        Object {
          "colId": "publishedAt",
          "dataType": "timestamp",
          "enableSearch": true,
          "field": "publishedAt",
          "headerName": "Publish date",
          "javaVariableName": "publishedAt",
          "minWidth": 150,
          "searchDataType": "regionalMidnight",
        },
        Object {
          "cellRenderer": "publishErrors",
          "colId": "publishErrors",
          "field": "publishErrors",
          "headerName": "Errors",
          "javaVariableName": "publishErrors",
          "minWidth": 150,
        },
      ]
    }
    onGridEvent={[Function]}
    rowSelection="single"
    suppressRowClickSelection={true}
  />
  <NexusDrawer
    direction="fromRight"
    headerContent={null}
    isClosedOnBlur={true}
    isLoading={false}
    isOpen={false}
    onClose={[Function]}
    title="Publish Errors"
    width="wider"
  >
    <div
      className="nexus-c-sync-log-table__errors-table"
    >
      <div
        className="nexus-c-sync-log-table__errors-table--header-cell"
        key="type"
      >
        TYPE
      </div>
      <div
        className="nexus-c-sync-log-table__errors-table--header-cell"
        key="description"
      >
        DESCRIPTION
      </div>
    </div>
  </NexusDrawer>
</div>
`;
